lib/updatehub/updatehub.c

```diff
@@ -0,0 +1,920 @@
+		goto error;
+	}
+
+	if (!start_coap_client()) {
+		LOG_ERR("Could not connect");
+		_ctx.code_status = UPDATEHUB_NETWORKING_ERROR;
+		goto error;
+	}
+
+	if (send_request(COAP_TYPE_CON, COAP_METHOD_POST, UPDATEHUB_PROBE) < 0) {
+		_ctx.code_status = UPDATEHUB_NETWORKING_ERROR;
+		goto cleanup;
+	}
+
+	memset(metadata, 0, MAX_PAYLOAD_SIZE);
+	probe_cb(metadata);
+
+	if (_ctx.code_status != UPDATEHUB_OK) {
+		goto cleanup;
+	}
+
+	memset(&_update_info, 0, sizeof(_update_info));
+	if (metadata_hash_get(metadata) < 0) {
+		LOG_ERR("Could not get metadata hash");
+		_ctx.code_status = UPDATEHUB_METADATA_ERROR;
+		goto cleanup;
+	}
+
+	struct resp_probe_some_boards metadata_some_boards;
+	struct resp_probe_any_boards metadata_any_boards;
+
+	memcpy(metadata_copy, metadata, strlen(metadata));
+	if (json_obj_parse(metadata, strlen(metadata),
+			   recv_probe_sh_array_descr,
+			   ARRAY_SIZE(recv_probe_sh_array_descr),
+			   &metadata_some_boards) < 0) {
+
+		if (json_obj_parse(metadata_copy, strlen(metadata_copy),
+				   recv_probe_sh_string_descr,
+				   ARRAY_SIZE(recv_probe_sh_string_descr),
+				   &metadata_any_boards) < 0) {
+			LOG_ERR("Could not parse json");
+			_ctx.code_status = UPDATEHUB_METADATA_ERROR;
+			goto cleanup;
+		}
+
+		memcpy(_update_info.sha256sum_image,
+		       metadata_any_boards.objects[1].objects.sha256sum,
+		       strlen(metadata_any_boards.objects[1].objects.sha256sum));
+		_update_info.image_size = metadata_any_boards.objects[1].objects.size;
+	} else {
+		if (!is_compatible_hardware(&metadata_some_boards)) {
+			LOG_ERR("Incompatible hardware");
+			_ctx.code_status =
+				UPDATEHUB_INCOMPATIBLE_HARDWARE;
+			goto cleanup;
+		}
+		memcpy(_update_info.sha256sum_image,
+		       metadata_some_boards.objects[1].objects.sha256sum,
+		       strlen(metadata_some_boards.objects[1]
+			      .objects.sha256sum));
+		_update_info.image_size =
+			metadata_some_boards.objects[1].objects.size;
+	}
+
+	_ctx.code_status = UPDATEHUB_HAS_UPDATE;
+
+cleanup:
+	cleanup_conection();
+
+error:
+	k_free(metadata);
+	k_free(metadata_copy);
+	k_free(firmware_version);
+	k_free(device_id);
+
+	return _ctx.code_status;
+}
+
+enum updatehub_response updatehub_update(void)
+{
+	if (report(UPDATEHUB_STATE_DOWNLOADING) < 0) {
+		LOG_ERR("Could not reporting downloading state");
+		goto error;
+	}
+
+	if (report(UPDATEHUB_STATE_INSTALLING) < 0) {
+		LOG_ERR("Could not reporting installing state");
+		goto error;
+	}
+
+	if (install_update() != UPDATEHUB_OK) {
+		goto error;
+	}
+
+	if (report(UPDATEHUB_STATE_DOWNLOADED) < 0) {
+		LOG_ERR("Could not reporting downloaded state");
+		goto error;
+	}
+
+	if (report(UPDATEHUB_STATE_INSTALLED) < 0) {
+		LOG_ERR("Could not reporting installed state");
+		goto error;
+	}
+
+	if (report(UPDATEHUB_STATE_REBOOTING) < 0) {
+		LOG_ERR("Could not reporting rebooting state");
+		goto error;
+	}
+
+	LOG_INF("Image flashed successfully, you can reboot now");
+
+#if defined(CONFIG_UPDATEHUB_LOG)
```

Comment:
What is the purpose of this while loop

Reply:
Just wait for the LOG_INF print on the screen.

Reply:
You do not need to do it like this. If you want synchronous prints, then just enable CONFIG_LOG_IMMEDIATE